pkg/operator/baremetal_config.go

```diff
@@ -22,68 +24,107 @@ const (
+		return BaremetalProvisioningConfig{}, errors.Wrap(err, errMsg)
 	}
-	provisioningDHCPRange, found, err := unstructured.NestedString(provisioningSpec, "provisioningDHCPRange")
-	if !found || err != nil {
-		glog.Errorf("provisioningDHCPRange not found in Baremetal provisioning CR %s", configName)
+	// If provisioningNetwork is not provided, set its value based on provisioningDHCPExternal
+	provisioningNetwork, foundNetworkState, err := unstructured.NestedString(provisioningSpec, "provisioningNetwork")
+	if err != nil {
+		errMsg := fmt.Sprintf("Error reading provisioningNetwork from Baremetal provisioning CR %s", configName)
+		return BaremetalProvisioningConfig{}, errors.Wrap(err, errMsg)
+	}
+	if !foundNetworkState {
+		// Check if provisioningDHCPExternal is present in the config
+		provisioningDHCPExternal, foundDHCP, err := unstructured.NestedBool(provisioningSpec, "provisioningDHCPExternal")
+		if !foundDHCP || err != nil {
+			// Both the new provisioningNetwork and the old provisioningDHCPExternal configs are not found.
+			errMsg := fmt.Sprintf("Both the provisioningNetwork and provisioningDHCPExternal are not found in Baremetal provisioning CR %s", configName)
+			return BaremetalProvisioningConfig{}, errors.Wrap(err, errMsg)
+		}
+		if !provisioningDHCPExternal {
+			provisioningNetwork = provisioningNetworkManaged
+		} else {
+			provisioningNetwork = provisioningNetworkUnmanaged
+		}
+	}
+
+	// provisioningNetworkCIDR needs to be present for all provisioningNetwork states (even Disabled).
+	// The CIDR of the network needs to be extracted to form the provisioningIPCIDR
+	provisioningNetworkCIDR, found, err := unstructured.NestedString(provisioningSpec, "provisioningNetworkCIDR")
+	if err != nil {
+		errMsg := fmt.Sprintf("Error reading provisioningNetworkCIDR in Baremetal provisioning CR %s", configName)
+		return BaremetalProvisioningConfig{}, errors.Wrap(err, errMsg)
+	}
+	if !found {
+		errMsg := fmt.Errorf("A valid Network CIDR needs to be provided in the Baremetal provisioning CR %s for all provisioning network types", configName)
+		return BaremetalProvisioningConfig{}, errMsg
+	}
+
+	// Check if the other config values make sense for the provisioningNetwork configured.
+	if provisioningInterface == "" && provisioningNetwork == provisioningNetworkManaged {
+		err := fmt.Errorf("provisioningInterface cannot be empty when provisioningNetwork is Managed.")
 		return BaremetalProvisioningConfig{}, err
 	}
-	provisioningOSDownloadURL, found, err := unstructured.NestedString(provisioningSpec, "provisioningOSDownloadURL")
-	if !found || err != nil {
-		glog.Errorf("provisioningOSDownloadURL not found in Baremetal provisioning CR %s", configName)
+	if provisioningDHCPRange == "" && provisioningNetwork == provisioningNetworkManaged {
+		fmt.Errorf("provisioningDHCPRange cannot be empty when provisioningNetwork is Managed or when the DHCP server needs to run with the metal3 cluster.")
```

Comment:
I think the govet error is because the result of this fmt.Errorf isn't assigned to err so we can return it below?

Reply:
Thanks for pointing that out!